flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java

```diff
@@ -15,75 +16,468 @@
+
+			final Iterator<Entry<String, IRichBolt>> boltsIterator = bolts.entrySet().iterator();
+			while (boltsIterator.hasNext()) {
+
+				final Entry<String, IRichBolt> bolt = boltsIterator.next();
+				final String boltId = bolt.getKey();
+				final IRichBolt userBolt = copyObject(bolt.getValue());
+
+				final ComponentCommon common = stormTopology.get_bolts().get(boltId).get_common();
+
+				Set<Entry<GlobalStreamId, Grouping>> unprocessedBoltInputs = unprocessdInputsPerBolt.get(boltId);
+				if (unprocessedBoltInputs == null) {
+					unprocessedBoltInputs = new HashSet<>();
+					unprocessedBoltInputs.addAll(common.get_inputs().entrySet());
+					unprocessdInputsPerBolt.put(boltId, unprocessedBoltInputs);
+				}
+
+				// check if all inputs are available
+				final int numberOfInputs = unprocessedBoltInputs.size();
+				int inputsAvailable = 0;
+				for (Entry<GlobalStreamId, Grouping> entry : unprocessedBoltInputs) {
+					final String producerId = entry.getKey().get_componentId();
+					final String streamId = entry.getKey().get_streamId();
+					final HashMap<String, DataStream<Tuple>> streams = availableInputs.get(producerId);
+					if (streams != null && streams.get(streamId) != null) {
+						inputsAvailable++;
+					}
+				}
+
+				if (inputsAvailable != numberOfInputs) {
+					// traverse other bolts first until inputs are available
+					continue;
+				} else {
+					makeProgress = true;
+					boltsIterator.remove();
+				}
+
+				final Map<GlobalStreamId, DataStream<Tuple>> inputStreams = new HashMap<>(numberOfInputs);
+
+				for (Entry<GlobalStreamId, Grouping> input : unprocessedBoltInputs) {
+					final GlobalStreamId streamId = input.getKey();
+					final Grouping grouping = input.getValue();
+
+					final String producerId = streamId.get_componentId();
+
+					final Map<String, DataStream<Tuple>> producer = availableInputs.get(producerId);
+
+					inputStreams.put(streamId, processInput(boltId, userBolt, streamId, grouping, producer));
+				}
+
+				final Iterator<Entry<GlobalStreamId, DataStream<Tuple>>> iterator = inputStreams.entrySet().iterator();
+
+				final Entry<GlobalStreamId, DataStream<Tuple>> firstInput = iterator.next();
+				GlobalStreamId streamId = firstInput.getKey();
+				DataStream<Tuple> inputStream = firstInput.getValue();
+
+				final SingleOutputStreamOperator<?, ?> outputStream;
+
+				switch (numberOfInputs) {
+					case 1:
+						outputStream = createOutput(boltId, userBolt, streamId, inputStream);
+						break;
+					case 2:
+						Entry<GlobalStreamId, DataStream<Tuple>> secondInput = iterator.next();
+						GlobalStreamId streamId2 = secondInput.getKey();
+						DataStream<Tuple> inputStream2 = secondInput.getValue();
+						outputStream = createOutput(boltId, userBolt, streamId, inputStream, streamId2, inputStream2);
+						break;
+					default:
+						throw new UnsupportedOperationException("Don't know how to translate a bolt "
+								+ boltId + " with " + numberOfInputs + " inputs.");
+				}
```

Comment:
This is an improvement. Before, it would just blindly take one input at a time and create an operator. As far as I understand, Flink only supports two inputs at a time. I would like to address any further improvements in this regard after the merge.